cmd/zrepl/zrepl.c

```diff
@@ -277,6 +277,45 @@ uzfs_zvol_socket_write(int fd, char *buf, int nbytes)
 	return (0);
 }
 
+/*
+ * We expect only one chunk of data with meta header in write request.
+ * Nevertheless the code is general to handle even more of them.
+ */
+static int
+uzfs_submit_writes(zvol_info_t *zinfo, zvol_io_cmd_t *zio_cmd)
+{
+	blk_metadata_t	metadata;
+	zvol_io_hdr_t 	*hdr = &zio_cmd->hdr;
+	struct zvol_io_rw_hdr *write_hdr;
+	char	*p = (char *)zio_cmd->buf;
+	size_t	remain = hdr->len;
+	int	rc = 0;
+
+	while (remain > 0) {
+		if (remain < sizeof (*write_hdr))
+			return (-1);
+
+		write_hdr = (struct zvol_io_rw_hdr *)p;
+		metadata.io_num = write_hdr->io_num;
+
+		p += sizeof (*write_hdr);
+		remain -= sizeof (*write_hdr);
+		if (remain < write_hdr->len)
+			return (-1);
+
+		rc = uzfs_write_data(zinfo->zv, p,
```

Comment:
rc = uzfs_write_data(zinfo->zv, p, hdr->offset + hdr->len - remain, write_hdr->len, &metadata, B_FALSE);
                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
hdr->offset + write_hdr->len should be enought to calculate write offset. No ?